### DISCUSSION TITLE
feature request: Use display mode katex for block equations

### DISCUSSION THREAD

**Discussion by weakit** [open-webui/open-webui#6457] [General]
Taken from #4008:

Block equations are still rendered as inline KaTex:
![370070796-21c226cc-5941-45ad-9789-3faee1274792](https://github.com/user-attachments/assets/dab688f6-b147-47e4-a730-78b10a3df111)

I think they should better be rendered as display mode KaTex, which is very much easier to read, like on ChatGPT:
![370070881-d116721a-dfe3-43d0-a378-1372b7b638c1](https://github.com/user-attachments/assets/249f41b7-23e3-40ca-bb30-0978781c9b23)

This should be a simple fix, I can submit a PR if necessary.

**Comment by tjc0726** [comment]
Hi, I'm encounter the same problem (but maybe different reaon?) Anyway, after spending sometime digging into this problem,I find out my problem is cuz of katex-extension.ts are implement it intentonally

> Block delimiters doubles as inline delimiters when not followed by a newline

and my model is output as

`$$formula$$`

instead of 

```
$$
formula
$$
```

Have no idea why openwebui implement that way , but here's a function I wrote as a workaround

```
from pydantic import BaseModel
from typing import Optional
import re


class Filter:
    class Valves(BaseModel):
        pass

    def __init__(self):
        self.valves = self.Valves()

    def outlet(self, body: dict) -> dict:
        for message in body.get("messages", []):
            if message.get("role") == "assistant":
                content = message.get("content", "")
                if content:
                    message["content"] = self.fix_katex_blocks(content)
        return body

    def fix_katex_blocks(self, text: str) -> str:
        pattern = r"\$\$([^\$]+?)\$\$"

        def replacer(match):
            formula = match.group(0)
            lines = formula.split("\n")
            if (
                len(lines) >= 2
                and lines[0].strip() == "$$"
                and lines[-1].strip() == "$$"
            ):
                return formula

            content = match.group(1).strip()
            return f"\n$$\n{content}\n$$\n"

        result = re.sub(pattern, replacer, text)

        result = re.sub(r"\n{3,}", "\n\n", result)

        return result

```
